### DISCUSSION TITLE
Where can I call NimBLEDevice::init()?

### DISCUSSION THREAD

**Discussion by mauricego** [h2zero/NimBLE-Arduino#476] [General]
Hello,

if I call `NimBLEDevice::init()` directly inside the `setup()` function everything works.

But if I put `NimBLEDevice::init()` into a function in another file and call that function form my `setup()` I get the following error:

> ESP_ERROR_CHECK failed: esp_err_t 0x101 (ESP_ERR_NO_MEM) at 0x40094dfc
> file: ".pio/libdeps/az-delivery-devkit-v4/NimBLE-Arduino/src/NimBLEDevice.cpp" line 879
> func: static void NimBLEDevice::init(const string&)
> expression: esp_bt_controller_init(&bt_cfg)
> 
> abort() was called at PC 0x40094dff on core 1
> 
> 
> Backtrace:0x40084029:0x3ffcd6400x40094e09:0x3ffcd660 0x4009a6e5:0x3ffcd680 0x40094dff:0x3ffcd700 0x400d869e:0x3ffcd720 0x400d5c92:0x3ffcd770 0x400d625f:0x3ffcd7d0 0x400d57a5:0x3ffcd840 0x400f923a:0x3ffcd870

Is it not possible to call `NimBLEDevice::init()` outside of `setup()`?

**Comment by h2zero** [comment]
You can call it anywhere you need. I suspect something is releasing the ble controller memory before init is called.

**Comment by mauricego** [comment]
Do you know in which kind of direction I should start looking to find out what could release the ble controller memory?

**Comment by chegewara** [comment]
I would say there is no heap to init bluetooth stack.
Thats true, you can call `init()` everywhere, but it is recommended to do it in `setup()` to make sure there is enough heap to init stack.

**Comment by mauricego** [comment]
Turns out is has nothing to do with NimBLE.

Before calling `NimBLEDevice::init()` i called the following function:

```
void createName() {
    uint8_t baseMac[6];
    // Get MAC address for WiFi station
    esp_read_mac(baseMac, ESP_MAC_WIFI_STA);
    // Write unique name into apName
    sprintf(apName, "LED-Strip-%02X%02X%02X%02X%02X%02X", baseMac[0], baseMac[1], baseMac[2], baseMac[3], baseMac[4],
            baseMac[5]);
}
```

I used this code before calling `NimBLEDevice::init()` to generate a unique name for my ESP32.

But somehow this function crashes the ESP with the error of my original post.

I will do some further testing because I don't understand why this happens at all.

Before using NimBLE I used the normal Bluetooth BLEDevice.h and had no problem with calling createName() before calling the Bluetooth init function.
